llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -1772,15 +1772,17 @@ void AArch64TargetLowering::addTypeForNEON(MVT VT) {
 
 bool AArch64TargetLowering::shouldExpandGetActiveLaneMask(EVT ResVT,
                                                           EVT OpVT) const {
-  // Only SVE has a 1:1 mapping from intrinsic -> instruction (whilelo).
-  if (!Subtarget->hasSVE())
+  // Only SVE/SME has a 1:1 mapping from intrinsic -> instruction (whilelo).
+  if (!Subtarget->hasSVEorSME())
```

Comment:
Can we add a run line for sve2p1 and/or  sme in
CodeGen/AArch64/active_lane_mask.ll
To make sure these changes did not break any other test:

```
+++ b/llvm/test/CodeGen/AArch64/active_lane_mask.ll
@@ -1,6 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve < %s | FileCheck %s
-
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve2p1 < %s | FileCheck %s
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2 < %s | FileCheck %s
```

Probably it will break for the tests nxv32i1 for the sve2p1, maybe only sme then.

Reply:
I agree with @CarolineConcatto. There are actually two changes happening here:

1. We are permitting the direct lowering of get.active.lane.mask intrinsic calls to while instructions when using SME, even if SVE is not present. It makes sense, but we should have an extra RUN line for SME in active_lane_mask.ll, i.e. -mattr=+sme because the nxv2i1, nxv4i1, nxv8i1 and nxv16i1 types should all work for SME.
2. We are introducing support for the nxv32i1 type when either SVE2p1 or SME2 is available. So we need two RUN lines in get-active-lane-mask-32x1.ll - one for SVE2p1 and one for SME2.

Reply:
Ack